### PR TITLE
accessibility improvements

### LINKED ISSUE #1590: VoiceOver navigation shows a blank line after each message
Is this an interface bug?

I navigate with VoiceOver and read messages. VoiceOver doesn't speak after jumping to the next or previous message. I have to swipe twice to jump to the previous or next one.

### LINKED ISSUE #1722: VoiceOver: Add unread messages to account switch accessibility label
Currently the amount of unread messages is read out by VoiceOver just as a number, the context of the number is missing. With #1721 VoiceOver doesn't read it out at all.

It would be nice to append a String "X unread messages" to the accessibility label of the account switch avatar, if there are any unread messages.

### PR DESCRIPTION
**Pull request by cyBerta**
~~read out unread messages in account switch button and account switch view controller, targets #1722~~ EDIT: corresponding commit moved to separate pr #1751

~~targets #1538~~ EDIT: moved to separate pr #1752

closes #1590

~~targets #1741~~ EDIT: moved to separate pr #1750

### CODE REVIEW

**Review by r10s — COMMENTED** [review]
i rebased the pr; first impression: good to sort labels and hints etc.

however, i am not convinced that changing the layout for accessibility is a good idea in general, this makes accessibility harder to maintain, and will easily degrade accessibility. also, UI guidelines say that if there is only one button, here "Done", this has to be on the _right_ side - changing that will make accessibility potentially harder.

but maybe i am also missing sth, i am still playing around, we should have an av on that.

EDIT: i moved the fix https://github.com/deltachat/deltachat-ios/pull/1738/commits/7b841d630910e3855b447202deb481b60c3666e8 to another pr and merged it, that way we can handle this one a bit more relaxed :)

File: deltachat-ios/Chat/ChatViewController.swift [hunk 1]
```diff
@@ -1843,6 +1842,28 @@ class ChatViewController: UITableViewController, UITableViewDropDelegate {
...
+            case #selector(BaseMessageCell.messageCopy(_:)):
+                copyItem.onPerform?(indexPath)
+            case #selector(BaseMessageCell.messageReply(_:)):
+                replyItem.onPerform?(indexPath)
+            case #selector(BaseMessageCell.messageReplyPrivately(_:)):
+                replyPrivatelyItem.onPerform?(indexPath)
+            case #selector(BaseMessageCell.profileSelected(_:)):
+                // this action is only available from VoiceOver, it's the same as avatarTapped for non-VoiceOver users
+                avatarTapped(indexPath: indexPath)
+            default: break
```
Comment (r10s) [inline deltachat-ios/Chat/ChatViewController.swift:1862]: how can that be tested?

File: deltachat-ios/Controller/AudioRecorderController.swift [hunk 2]
```diff
@@ -112,9 +115,14 @@ class AudioRecorderController: UIViewController, AVAudioRecorderDelegate {
...
 
-        self.navigationItem.title = String.localized("voice_message")
-        self.navigationItem.leftBarButtonItem = cancelButton
-        self.navigationItem.rightBarButtonItem = doneButton
+        if UIAccessibility.isVoiceOverRunning {
+            self.navigationItem.leftBarButtonItem = doneButton
+        } else {
+            self.navigationItem.title = String.localized("voice_message")
+            self.navigationItem.leftBarButtonItem = cancelButton
+            self.navigationItem.rightBarButtonItem = doneButton
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply deltachat-ios/Controller/AudioRecorderController.swift:123]: so lets's revert that piece.

File: deltachat-ios/Controller/AudioRecorderController.swift [hunk 3]
```diff
@@ -46,11 +46,14 @@ class AudioRecorderController: UIViewController, AVAudioRecorderDelegate {
         view.translatesAutoresizingMaskIntoConstraints = false
         view.alpha = 0.0
         view.contentMode = UIView.ContentMode.scaleAspectFit
+        view.isAccessibilityElement = true
```
Comment (cyBerta) [inline deltachat-ios/Controller/AudioRecorderController.swift:49]: would be nice to get that piece  of code in though. I can also do a separate PR for it.

File: deltachat-ios/Controller/AudioRecorderController.swift [hunk 4]
```diff
@@ -152,13 +160,6 @@ class AudioRecorderController: UIViewController, AVAudioRecorderDelegate {
         validateMicrophoneAccess()
     }
 
-    override func viewDidAppear(_ animated: Bool) {
```
Comment (cyBerta) [inline deltachat-ios/Controller/AudioRecorderController.swift:155]: that needs to be reverted as well then

File: deltachat-ios/Controller/AudioRecorderController.swift [hunk 5]
```diff
@@ -292,6 +295,10 @@ class AudioRecorderController: UIViewController, AVAudioRecorderDelegate {
                     self.waveFormView.alpha = granted ? 1.0 : 0.0
                     self.doneButton.isEnabled = granted
 
+                    if UIAccessibility.isVoiceOverRunning && !granted {
```
Comment (cyBerta) [inline deltachat-ios/Controller/AudioRecorderController.swift:298]: revert

File: deltachat-ios/Chat/ChatViewController.swift (hunk 1 above)
Comment (r10s) [reply deltachat-ios/Chat/ChatViewController.swift:1862]: > how can that be tested?

after av with @cyBerta: swipe up/down in voiceOver mode should read the menu.

**Review by r10s — APPROVED** [review]

**Pull request merged**